File: examples/demux.rs
use std::env;
use std::fs::File;
use std::io::{self, Read, Seek, SeekFrom};
use std::path::Path;

use shiguredo_mp4::demux::{Input, Mp4FileDemuxer};

const BUFFER_SIZE: usize = 1024 * 1024; // 1MB のバッファサイズ

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <mp4_file>", args[0]);
        std::process::exit(1);
    }

    let filepath = &args[1];
    if !Path::new(filepath).exists() {
        eprintln!("Error: Could not open file '{}'", filepath);
        std::process::exit(1);
    }

    // ファイルを開く（メモリ全体には読み込まない）
    let mut file = File::open(filepath)?;

    // デマルチプレックス処理を初期化
    let mut demuxer = Mp4FileDemuxer::new();
    let mut buffer = vec![0u8; BUFFER_SIZE];

    // 初期化完了まで必要なデータを段階的に読み込む
    while let Some(required) = demuxer.required_input() {
        // 必要なデータを読み込む
        let data = read_from_file(&mut file, required.position, required.size, &mut buffer)?;
        let input = Input {
            position: required.position,
            data,
        };
        demuxer.handle_input(input);
    }

    // トラック情報を取得する
    let tracks = match demuxer.tracks() {
        Ok(tracks) => tracks,
        Err(e) => {
            eprintln!("Error: Failed to get tracks: {}", e);
            std::process::exit(1);
        }
    };

    println!("Found {} track(s)\n", tracks.len());

    // トラック情報を表示
    for (i, track) in tracks.iter().enumerate() {
        println!("Track {}:", i + 1);
        println!("  Track ID: {}", track.track_id);
        println!(
            "  Duration: {} (timescale: {})",
            track.duration, track.timescale
        );
        println!();
    }

    // サンプル情報を表示
    let mut sample_count = 0;
    println!("Samples:");

    // 時系列順にサンプルを抽出する
    loop {
        match demuxer.next_sample() {
            Ok(Some(sample)) => {
                sample_count += 1;
                println!("  Sample {}:", sample_count);
                println!("    Track ID: {}", sample.track.track_id);
                println!("    Timestamp: {}", sample.timestamp);
                println!("    Data offset: 0x{:x}", sample.data_offset);
                println!("    Data size: {} bytes", sample.data_size);
                println!();

                if sample_count >= 10 {
                    println!("  ... (showing first 10 samples)");
                    break;
                }
            }
            Ok(None) => break,
            Err(e) => {
                eprintln!("Error: {}", e);
                std::process::exit(1);
            }
        }
    }

    Ok(())
}

/// ファイルの指定位置から必要なサイズだけデータを読み込む
fn read_from_file<'a>(
    file: &mut File,
    position: u64,
    size: Option<usize>,
    buffer: &'a mut [u8],
) -> io::Result<&'a [u8]> {
    file.seek(SeekFrom::Start(position))?;

    let read_size = match size {
        Some(s) => s.min(buffer.len()),
        None => buffer.len(),
    };

    let bytes_read = file.read(&mut buffer[..read_size])?;
    Ok(&buffer[..bytes_read])
}

File: crates/wasm/examples/demux.js
/**
 * Node.js で mp4-rust の WASM を使って MP4 ファイルをデマルチプレックスする例
 *
 * # 使用方法
 *
 * ```bash
 * node demux.js /path/to/input.mp4
 * ```
 */

import fs from 'fs';
import path from 'path';

// WASM モジュールを動的にインポート
const wasmModule = await import('../target/wasm32-unknown-unknown/release/mp4_wasm.wasm', {
    with: { type: 'module' }
});

const {
    memory,
    mp4_file_demuxer_new,
    mp4_file_demuxer_free,
    mp4_file_demuxer_get_required_input,
    mp4_file_demuxer_handle_input,
    mp4_file_demuxer_get_tracks,
    mp4_file_demuxer_next_sample,
    mp4_file_demuxer_get_last_error,
    mp4_demux_track_info_to_json,
    mp4_demux_sample_to_json,
    mp4_vec_ptr,
    mp4_vec_len,
    mp4_vec_free,
} = wasmModule;

// メモリバッファへのアクセスを簡略化する関数群
function readCString(ptr) {
    const view = new Uint8Array(memory.buffer, ptr);
    let length = 0;
    while (view[length] !== 0) length++;
    return new TextDecoder().decode(view.slice(0, length));
}

function readU8Array(ptr, len) {
    return new Uint8Array(memory.buffer, ptr, len);
}

function readJSON(vecPtr) {
    const ptr = mp4_vec_ptr(vecPtr);
    const len = mp4_vec_len(vecPtr);
    const bytes = readU8Array(ptr, len);
    const jsonStr = new TextDecoder().decode(bytes);
    mp4_vec_free(vecPtr);
    return JSON.parse(jsonStr);
}

async function demuxMP4File(filePath) {
    if (!fs.existsSync(filePath)) {
        console.error(`Error: File not found: ${filePath}`);
        process.exit(1);
    }

    const fileSize = fs.statSync(filePath).size;
    const file = fs.openSync(filePath, 'r');

    // デマルチプレックサーを作成
    const demuxerPtr = mp4_file_demuxer_new();
    if (!demuxerPtr) {
        console.error('Error: Could not create demuxer');
        process.exit(1);
    }

    // ファイルデータを読み込んでデマルチプレックス処理を進める
    console.log(`Demuxing ${path.basename(filePath)}...`);

    while (true) {
        // ポインタを確保（読み書き用）
        const requiredPosPtr = mp4_alloc(8);  // uint64_t
        const requiredSizePtr = mp4_alloc(4); // int32_t

        // 次に必要な入力データの位置とサイズを取得
        const err = mp4_file_demuxer_get_required_input(
            demuxerPtr,
            requiredPosPtr,
            requiredSizePtr
        );

        if (err !== 0) { // MP4_ERROR_OK
            const errorMsg = readCString(mp4_file_demuxer_get_last_error(demuxerPtr));
            console.error(`Error: Failed to get required input: ${errorMsg}`);
            mp4_file_demuxer_free(demuxerPtr);
            fs.closeSync(file);
            process.exit(1);
        }

        // 値を読み取る
        const requiredPosView = new BigUint64Array(memory.buffer, requiredPosPtr, 1);
        const requiredSizeView = new Int32Array(memory.buffer, requiredSizePtr, 1);

        const requiredPos = requiredPosView[0];
        let requiredSize = requiredSizeView[0];

        mp4_free(requiredPosPtr, 8);
        mp4_free(requiredSizePtr, 4);

        // 必要なデータサイズが 0 の場合は初期化完了
        if (requiredSize === 0) {
            break;
        }

        // 読み込むサイズを決定
        let readSize = requiredSize;
        if (requiredSize === -1) {
            // ファイル末尾までの読み込みが必要
            readSize = fileSize - Number(requiredPos);
        }

        // ファイルからデータを読み込む
        const buffer = Buffer.alloc(readSize);
        fs.readSync(file, buffer, 0, readSize, Number(requiredPos));

        // WASM メモリにコピー
        const wasmBuffer = mp4_alloc(readSize);
        const wasmView = new Uint8Array(memory.buffer, wasmBuffer, readSize);
        wasmView.set(buffer);

        // デマルチプレックサーにデータを入力
        mp4_file_demuxer_handle_input(
            demuxerPtr,
            requiredPos,
            wasmBuffer,
            readSize
        );

        mp4_free(wasmBuffer, readSize);
    }

    // トラック情報を取得
    const tracksPtr = mp4_alloc(8); // ポインタ
    const trackCountPtr = mp4_alloc(4); // uint32_t

    const err = mp4_file_demuxer_get_tracks(
        demuxerPtr,
        tracksPtr,
        trackCountPtr
    );

    if (err !== 0) {
        const errorMsg = readCString(mp4_file_demuxer_get_last_error(demuxerPtr));
        console.error(`Error: Failed to get tracks: ${errorMsg}`);
        mp4_file_demuxer_free(demuxerPtr);
        fs.closeSync(file);
        process.exit(1);
    }

    // トラック数を取得
    const trackCountView = new Uint32Array(memory.buffer, trackCountPtr, 1);
    const trackCount = trackCountView[0];

    console.log(`\nFound ${trackCount} track(s)\n`);

    // トラック情報を表示
    const tracksPointerView = new BigUint64Array(memory.buffer, tracksPtr, 1);
    const tracksPointer = Number(tracksPointerView[0]);

    for (let i = 0; i < trackCount; i++) {
        // Mp4DemuxTrackInfo のサイズを計算（id: u32, kind: u32, duration: u64, timescale: u32）
        const trackInfoSize = 4 + 4 + 8 + 4; // = 20 bytes
        const trackInfoPtr = tracksPointer + i * trackInfoSize;

        const jsonPtr = mp4_demux_track_info_to_json(trackInfoPtr);
        const trackInfo = readJSON(jsonPtr);

        console.log(`Track ${i + 1}:`);
        console.log(`  Track ID: ${trackInfo.track_id}`);
        console.log(`  Kind: ${trackInfo.kind}`);
        console.log(`  Duration: ${trackInfo.duration} (timescale: ${trackInfo.timescale})`);
        console.log(`  Duration (seconds): ${Number(trackInfo.duration) / trackInfo.timescale}`);
        console.log();
    }

    mp4_free(tracksPtr, 8);
    mp4_free(trackCountPtr, 4);

    // サンプル情報を取得・表示
    console.log('Samples:');
    let sampleCount = 0;
    let keyframeCount = 0;

    const samplePtr = mp4_alloc(256); // Mp4DemuxSample の最大サイズ

    while (sampleCount < 10) {
        const err = mp4_file_demuxer_next_sample(demuxerPtr, samplePtr);

        if (err === 8) { // MP4_ERROR_NO_MORE_SAMPLES
            break;
        }

        if (err !== 0) {
            const errorMsg = readCString(mp4_file_demuxer_get_last_error(demuxerPtr));
            console.error(`Error: Failed to get next sample: ${errorMsg}`);
            break;
        }

        sampleCount++;

        const jsonPtr = mp4_demux_sample_to_json(samplePtr);
        const sample = readJSON(jsonPtr);

        console.log(`  Sample ${sampleCount}:`);
        console.log(`    Track ID: ${sample.track_id}`);
        console.log(`    Keyframe: ${sample.keyframe ? 'Yes' : 'No'}`);
        console.log(`    Timestamp: ${sample.timestamp}`);
        console.log(`    Duration: ${sample.duration}`);
        console.log(`    Data offset: 0x${sample.data_offset.toString(16)}`);
        console.log(`    Data size: ${sample.data_size} bytes`);

        if (sample.sample_entry) {
            console.log(`    Codec: ${sample.sample_entry.kind}`);
        }

        console.log();

        if (sample.keyframe) {
            keyframeCount++;
        }
    }

    if (sampleCount >= 10) {
        console.log(`  ... (showing first 10 samples)\n`);
    }

    console.log(`Total samples processed: ${sampleCount}, Keyframes: ${keyframeCount}`);

    mp4_free(samplePtr, 256);

    // リソースを解放
    mp4_file_demuxer_free(demuxerPtr);
    fs.closeSync(file);
}

// メイン処理
const args = process.argv.slice(2);
if (args.length === 0) {
    console.error('Usage: node demux.js <mp4_file>');
    process.exit(1);
}

const filePath = args[0];
await demuxMP4File(filePath);

